### DISCUSSION TITLE
How can I add Mockito to the test classpath in Tycho's unit-tests with eclipse-plugin packaging

### DISCUSSION THREAD

**Discussion by jgsuess** [eclipse-tycho/tycho#386] [Q&A] [answered]
Recently, it has [become possible to execute unit-tests within maven projects that have `eclipse-plugin` packaging](https://bugs.eclipse.org/bugs/show_bug.cgi?id=572420). And, in addition there is [support for resolving JUnit Classpath Containers](https://bugs.eclipse.org/bugs/show_bug.cgi?id=572602).

I would like to execute unit-tests within projects with *eclipse-plugin* packaging, but would like to use the *Mockito library in addition to JUnit*. I have a [pomless build](https://wiki.eclipse.org/Tycho/pomless) and would like to keep it that way. I do not want to add non-PDE files to the build, unless this is unavoidable.

**Question**: What is the idiomatic/intended/correct way to add this dependency, or any other test-time dependencies?

I have tried `build.properties/additional.bundles`,  and `build.properties/jars.extra.classpath` with no result. I have not found any documentation on how to add test dependencies apart from the JUnit test container.

*Note*: I am aware of the use of fragments for unit testing. This is not what I am after. I actually want to use the new mechanism, if possible, or hear that this is currently impossible.

For my initial purposes, and given these are intended to be Unit-tests, running non-OSGI would be ok. If there is a means for OSGI as well, that would be great, but I cannot imagine where the platform configuration could be stored.

Disclaimer: This question has also been posted as [SO 69611036](https://stackoverflow.com/questions/69611036/how-can-i-add-mockito-to-the-test-classpath-in-tychos-unit-tests-with-eclipse-p).

Best,

JGS

**Comment by laeubi** [comment, marked as answer]
Sadly most of my attempts to make this happen in a convenient way where rejected by the PDE team but you still has the following options:

- you can add Mockito as an optional bundle dependency
- you can add a [M2_REPO Classpath variable reference](https://github.com/eclipse/tycho/pull/207)

**Reply by jgsuess** [reply]
Thanks so much! That should help a great deal. 

Just an idea: Given that M2_REPO is associated with M2 layouts, could we use [Dynamic Variables](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/extension-points/org_eclipse_core_variables_dynamicVariables.html), regexes, globs or something to make the version selection a bit more robust? What about adding our own `tycho.unit.test.bundles` key to `build.properties` using `platform:plugin/` syntax and then adding the resolved artifacts that overlap with `additional.bundles`.

I.e. if it is in `additional.bundles` and it is in `tycho.unit.test.bundles` add it to the surefire unit test dependencies.

**Reply by laeubi** [reply]
Currently `M2_REPO` support is very basic, but if you have a certain feature in mind don't hesitate to open a feature request (but don't forget to include an example) of the missing pieces.

Adding custom properties always has the drawback of only working on the command-line but not in the IDE...
